src/main/java/com/epam/aidial/core/upstream/UpstreamRouteProvider.java

```diff
@@ -0,0 +1,111 @@
+package com.epam.aidial.core.upstream;
+
+import com.epam.aidial.core.config.Addon;
+import com.epam.aidial.core.config.Application;
+import com.epam.aidial.core.config.Assistant;
+import com.epam.aidial.core.config.Config;
+import com.epam.aidial.core.config.Deployment;
+import com.epam.aidial.core.config.Model;
+import com.epam.aidial.core.config.Route;
+import com.epam.aidial.core.config.Upstream;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@Slf4j
+public class UpstreamRouteProvider {
```

Comment:
Please javadoc. It's worth mentioning that if upstream is changed due to config change then the balancer state is reset.